odoo/models.py

```diff
@@ -4269,6 +4255,78 @@ def _check_company(self, fnames=None):
                 })
             raise UserError("\n".join(lines))
 
+    def check_access(self, operation: str) -> None:
+        """ Verify that the current user is allowed to perform ``operation`` on
+        all the records in ``self``. The method raises an :class:`AccessError`
+        if the operation is forbidden on the model in general, or on any record
+        in ``self``.
+
+        In particular, when ``self`` is empty, the method checks whether the
+        current user has some permission to perform ``operation`` on the model
+        in general::
+
+            # check that user has some minimal permission on the model
+            records.browse().check_access(operation)
+
+        """
+        if not self.env.su:
+            self._check_access(operation, raise_exception=True)
+
+    def has_access(self, operation: str) -> bool:
+        """ Return whether the current user is allowed to perform ``operation``
+        on all the records in ``self``. The method is fully consistent with
+        method :meth:`check_access` but returns a boolean instead.
+        """
+        if not self.env.su and (forbidden := self._check_access(operation)):
+            forbidden.invalidate_recordset()
+            return False
+        return True
+
+    def _filter_access(self, operation: str):
+        """ Return the subset of ``self`` for which the current user is allowed
+        to perform ``operation``. The method is fully equivalent to::
+
+            self.filtered(lambda record: record.has_access(operation))
+
+        """
+        if self and not self.env.su and (forbidden := self._check_access(operation)):
+            forbidden.invalidate_recordset()
+            return self - forbidden
```

Comment:
This makes the model access check O(n) instead of O(1) :thinking: 
But maybe we consider that this is not a common case?

Reply:
I will make the code behave as before.
Thanks for pointing that out.
And the invalidation only makes sense when you don't have _read_ access, by the way.